pkg/authentication/externalidp/awsiam/aws_client.go

```diff
@@ -0,0 +1,153 @@
+package awsiam
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+	"github.com/google/uuid"
+)
+
+var ErrInvalidCredentialsFormat = errors.New("missing required parts in credentials")
+
+const (
+	AuthVersion       = "2011-06-15"
+	AuthMethod        = http.MethodPost
+	AuthAction        = "GetCallerIdentity"
+	AuthAlgorithm     = "AWS4-HMAC-SHA256"
+	StsGlobalEndpoint = "sts.amazonaws.com"
+	AuthActionKey     = "Action"
+	AuthVersionKey    = "Version"
+	AuthAlgorithmKey  = "X-Amz-Algorithm"
+	//nolint:gosec
+	AuthCredentialKey = "X-Amz-Credential"
+	AuthDateKey       = "X-Amz-Date"
+	AuthExpiresKey    = "X-Amz-Expires"
+	//nolint:gosec
+	AuthSecurityTokenKey  = "X-Amz-Security-Token"
+	AuthSignedHeadersKey  = "X-Amz-SignedHeaders"
+	AuthSignatureKey      = "X-Amz-Signature"
+	DatetimeFormat        = "20060102T150405Z"
+	CredentialTimeFormat  = "20060102"
+	DefaultSTSLoginExpire = 15 * time.Minute
+
+	mimLengthSplitCreds = 3
+)
+
+type AWSIdentityTokenInfo struct {
+	Method             string   `json:"method"`
+	Host               string   `json:"host"`
+	Region             string   `json:"region"`
+	Action             string   `json:"action"`
+	Date               string   `json:"date"`
+	ExpirationDuration string   `json:"expiration_duration"`
+	AccessKeyID        string   `json:"access_key_id"`
+	Signature          string   `json:"signature"`
+	SignedHeaders      []string `json:"signed_headers"`
+	Version            string   `json:"version"`
+	Algorithm          string   `json:"algorithm"`
+	SecurityToken      string   `json:"security_token"`
+}
+
+type IAMAuthParams struct {
+	ProviderType        string
+	TokenRequestHeaders map[string]string
+	URLPresignTTL       time.Duration
+	TokenTTL            time.Duration
+	RefreshInterval     time.Duration
+}
+
+func GenerateIdentityTokenInfo(ctx context.Context, params *IAMAuthParams, stsClient *sts.Client) (*AWSIdentityTokenInfo, error) {
+	url, err := PresignGetCallerIdentityFromAuthParams(ctx, params, stsClient)
+	if err != nil {
+		return nil, fmt.Errorf("generating sts presigned url: %w", err)
+	}
+	tokenInfo, err := ParsePresignedURL(url)
+	if err != nil {
+		return nil, fmt.Errorf("parsing credentials sts presigned url: %w", err)
+	}
+	return tokenInfo, nil
+}
+
+func ParsePresignedURL(presignedURL string) (*AWSIdentityTokenInfo, error) {
+	parsedURL, err := url.Parse(presignedURL)
+	if err != nil {
+		return nil, err
+	}
+
+	queryParams := parsedURL.Query()
+	credentials := queryParams.Get(AuthCredentialKey)
+	splitedCreds := strings.Split(credentials, "/")
+	if len(splitedCreds) < mimLengthSplitCreds {
+		return nil, fmt.Errorf("invalid credentials format: %w", ErrInvalidCredentialsFormat)
+	}
+	calculatedRegion := splitedCreds[2]
```

Comment:
Be consistent with access to splittedCreds, either set both `splitedCreds[2]` and `splitedCreds[0]` into dedicated variables, or both directly into `AWSIdentityTokenInfo` 

```go
&AWSIdentityTokenInfo{
    // ...
    Region: splittedCreds[2],
    AccessKeyID: splitedCreds[0],
    // ...
```